src/Magento/FunctionalTestingFramework/Extension/TestContextExtension.php

```diff
@@ -150,6 +115,26 @@ public function extractContext($trace, $class)
         return null;
     }
 
+    /**
+     * Attach suppressed exception thrown before _after hook to the current step.
+     * @param  \Exception $exception
+     * @return mixed
+     */
+    public function logPreviousException(\Exception $exception)
```

Comment:
Question:
* Should we attach every exception, *not* just the buried one? It feels strange to see a test fail in two steps but only see one `Exception` attached. We could even label it `Before/Test/After Exception` using some functions in this class already.

Reply:
Those were my initial thoughts too, but looks like `After` block's exception stack trace appears in the allure report twice - one at the top and another at the bottom after test steps. Will it be a duplication in some way? Capturing context and appending it to screenshot would work, however, there will only be 1 exception in `Before + Test body` scope, would context  be redundant to display? Also looks like Allure doesn't a clear separation of `Before/Test/After` blocks  of the test. What are your thoughts?

Reply:
So good news is that with just Codeception 2.4.0 we can do a couple things:
* Stop iterating over errors on line 88, they will only now be parent->previous() relationship
* Given above, we can instead iterate by doing logError($error), wherein we log the parent, check for child and call logError($childError) to pop the full stack
* remove the `testFail()` hook entirely, just use `testEnd()` to log these

We can modify the hook to look something like:

```
...
$testResultObject = call_user_func(\Closure::bind(
    function () use ($cest) {
        return $cest->getTestResultObject();
    },
    $cest
));
if (!empty($testResultObject->errors())) {
    foreach ($testResultObject->errors() as $error) {
        $this->attachExceptionToAllure($error->thrownException());
    }
}
if (!empty($testResultObject->failures())) {
    foreach ($testResultObject->failures() as $error) {
        $this->attachExceptionToAllure($error->thrownException());
    }
}
...
```

We can end up with something like this
![image](https://user-images.githubusercontent.com/28712019/69091107-3a78bd80-0a0f-11ea-8ade-620c048bc3fa.png)

You could add to the signature of `attachExceptionToAllure` to inlcude a prefix like `Error` vs `Failure` as well for clearer deliniation.

Reply:
Even if we have some duplication, I think if we lean on prefixing the attachments it would be best in case someone needs the info. Those can be ignored by simply not expanding them.

Reply:
Added context as a prefix to the attachment. Including a prefix like `Error` or `Failure` may need more investigation. From what I understood, `PHPUnit\Framework\Exception` are marked as failures and `Magento\Framework\Exception` are marked as Errors. Maybe there are exceptions which don't follow this criteria.
<img width="370" alt="Screen Shot 2019-11-19 at 9 22 53 AM" src="https://user-images.githubusercontent.com/8500612/69160010-6bf49600-0aae-11ea-8194-036f46a021dc.png">

Reply:
I think you are right in not trusting that the exception types are necessarily linked to failure vs error. You could include them in the signature of your new `attachExceptionToAllure` but I think having the prefix is enough for now.